benchmarks/src/jmh/java/org/postgresql/benchmark/stream/StreamReceiveInteger.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2023, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.benchmark.stream;
+
+import org.postgresql.core.VisibleBufferedInputStream;
+
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.infra.Blackhole;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+
+import java.io.ByteArrayInputStream;
+import java.util.concurrent.TimeUnit;
+
+@Fork(value = 5, jvmArgsPrepend = "-Xmx128m")
+@Measurement(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
+@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
+@State(Scope.Thread)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+public class StreamReceiveInteger {
+
+  private byte[] integerData;
+
+  @Setup
+  public void setUp() throws Exception {
+    // we only use a fixed buffer size of 8192 (8 KB)
+    integerData = new byte[8192];
+
+    for (int i = 0; i < integerData.length; i += 2) {
+      int value = i / 2;
+      integerData[i] = (byte) (value >> 8);
+      integerData[i + 1] = (byte) (value & 0xFF);
+    }
+
+  }
+
+  @Benchmark
+  public void receiveInteger4Benchmark(Blackhole bh) throws Exception {
+    VisibleBufferedInputStream bufferedInputStream = new VisibleBufferedInputStream(new ByteArrayInputStream(integerData),
+        integerData.length);
+
+    while (bufferedInputStream.available() >= 4) {
+      int value = bufferedInputStream.receiveInteger4();
+      bh.consume(value);
+    }
+  }
+
+  @Benchmark
+  public void receiveInteger2Benchmark(Blackhole bh) throws Exception {
+    VisibleBufferedInputStream bufferedInputStream = new VisibleBufferedInputStream(new ByteArrayInputStream(integerData),
+        integerData.length);
```

Comment:
This looks like a setup to me. Have you considered moving it to `@Setup` method?

Reply:
The suggestion of moving stream initialization to the setup was invalid as we can't reuse the stream, and we really have to recreate it otherwise it gets exhausted and we can no longer read from it.